docs/architecture.md

```diff
@@ -0,0 +1,216 @@
+variable respectively (3)).
+
+The `Get|Store|EraseCommand`s call the corresponding `Get|Store|EraseCredentialAsync` methods on the `IHostProvider`, passing the request from Git
+together with an instance of the `ICommandContext` (4). The host provider can then make use of various services available on the command context to
+complete the requested operation (5).
+
+Once a credential has been created, retrieved, stored or erased, the host provider returns the credential (for `get` operations only) to the calling
+command (6). The credential is then serialized and returned to Git over standard output (7) and GCM Core terminates with a successful exit code.
+
+## Host provider
+
+Host providers implement the `IHostProvider` interface. They can choose to directly implement the interface they can also derive from the `HostProvider`
+abstract class (which itself implements the `IHostProvider` interface).
+
+The `HostProvider` abstract class implements the `Get|Store|EraseCredentialAsync` methods and instead has a `GenerateCredentialAsync` and
+`GetCredentialKey` abstract methods. Calls to `get`, `store`, or `erase` result in first a call to `GetCredentialKey` which should return a stable and
+unique "key" for the request. This forms the key for any stored credential in the credential store. During a `get` operation the credential store is queried
+for an existing credential with the computed key. If a credential is found it is returned immediately. Similarly, calls to `store` and `erase` are handles
+automatically to store credentials against, and erase credentials matching the computed key. Methods are implemented as `virtual` meaning you can always
+override this behaviour, for example to clear other custom caches on an `erase` request, without having to reimplement the lookup/store credential logic.
+
+Host providers are queried in turn (registration order) via the `IHostProvider.IsSupported` method and passed the input recieved from Git.
+If the provider recognises the request, for example by a matching known host name, they can return `true`. If the provider wants to cancel and abort
+an authentication request, for example if this is a HTTP (not HTTPS) request for a known host, they should still return `true` and later cancel the request.
+
+Depending on the request from Git, one of `GetCredentialAsync` (for `get` requests), `StoreCredentialAsync` (for `store` requests) or
+`EraseCredentialAsync` (for `erase` requests) will be called. The argument `InputArguments` contains the request information passed over standard input
+from Git/the caller; the same as was passed to `IsSupported`.
+
+The return value for the `get` operation must be an `ICredential` that Git can use to complete authentication. The credential can also be an instance where
+both username and password are the empty string, to signal to Git it should let cURL use "any auth" detection - typically to use Windows Integrated Authentication.
```

Comment:
Having been burned by the double empty string thing with gvfs-helper recently, I'm wondering if this point should get a sub-heading or something.  :-) :-)

Reply:
or a footnote or link to curl's docs.